### PR TITLE
NEXUS-8305: Improvements required to main work (Part 1)

### PR DESCRIPTION
**Pull request by cstamas**
Splitting original PR https://github.com/sonatype/nexus-oss/pull/1257 into multiple smaller PR "steps" for easier review. This is step 1, with changes:
* improve MavenPath to contain parsed timestamp and build number
* improve Maven2MavenPathParser to parse the new fields above
* add (unused in this step) various constants to Maven2Format, to be used in next step
* make MavenPath, MavenPathParser publicly available, make DigestExtractor usable outside of it's package (to be used in next step)

Issue
https://issues.sonatype.org/browse/NEXUS-8305

CI
http://bamboo.s/browse/NX3-OSSF309

### CODE REVIEW

**Comment by cstamas** [comment]
http://bamboo.s/browse/NX3-OSSF309-1 :+1:

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/Maven2MavenPathParser.java [hunk 1]
```diff
@@ -96,6 +96,8 @@ else if (str.endsWith("." + HashType.MD5.getExt())) {
       }
 
       String version = baseVersion;
+      Long timestamp = null;
+      Integer buildNumber = null;
       String tail;
       if (snapshot) {
         int vSnapshotStart = artifactId.length() + baseVersion.length() - 9 + 2;
```
Comment (mrprescott) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/Maven2MavenPathParser.java]: This should be -10 + 3

File: plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/Maven2MavenPathParser.java [hunk 2]
```diff
@@ -106,23 +108,41 @@ else if (str.endsWith("." + HashType.MD5.getExt())) {
           tail = str.substring(nTailPos);
         }
         else {
-          StringBuilder snapshotBuildNumber = new StringBuilder(version);
-          snapshotBuildNumber
-              .append(str.substring(vSnapshotStart + version.length(), vSnapshotStart + version.length() + 8));
+          final StringBuilder snapshotTimestampedVersion = new StringBuilder(version);
```
Comment (mrprescott) [inline plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/Maven2MavenPathParser.java]: This method's getting a bit complicated, be nice to tease it apart. There's also a lot of magic numbers.

**Comment by mrprescott** [comment]
+1 with minor comments

**Comment by cstamas** [comment]
Merged as 58fabb9ba45972ce4db3123151dec00dfeae4085 (with applied cmments)

**Pull request closed without merging**